File: 2800-minimum-string-length-after-removing-substrings/2800-minimum-string-length-after-removing-substrings.cpp
class Solution {
public:
    int minLength(string s) {
        //same keep 2 pointers 
        int top = -1;
        for(int i = 0; i < s.size(); i++){
            if(top == -1){//empty stack
                s[++top] = s[i]; 
            }
            else if (s[top] == 'A' && s[i] == 'B' || s[top] == 'C' && s[i] == 'D'){
                top--;
            }
            else{
                s[++top] = s[i];
            }
        }
        return top + 1;
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();

File: 2800-minimum-string-length-after-removing-substrings/2800-minimum-string-length-after-removing-substrings.java
class Solution {
    public int minLength(String s) {
        Stack<Character> stack = new Stack<>();
        for(char c : s.toCharArray()){
            if(stack.isEmpty()){
                stack.push(c);
                continue;
            }
            switch(c){
                case 'B':
                    if(stack.peek() == 'A'){
                        stack.pop();
                    }
                    else{
                        stack.push(c);
                    }
                    break;
                case 'D':
                    if(stack.peek() == 'C'){
                        stack.pop();
                    }
                    else{
                        stack.push(c);
                    }
                    break;
                default:
                    stack.push(c);
                
            }
        }
        return stack.size();
    }
}